### PR TITLE
Improve apply_type_nothrow precision

### PR DESCRIPTION
**Pull request by Keno**

### CODE REVIEW

File: base/compiler/tfuncs.jl [hunk 1]
```diff
@@ -1082,22 +1096,37 @@ function apply_type_nothrow(argtypes::Array{Any, 1}, @nospecialize(rt))
...
-        if ai ⊑ TypeVar
+        if ai ⊑ TypeVar || ai === DataType
             # We don't know anything about the bounds of this typevar, but as
             # long as the UnionAll is not constrained, that's ok.
             if !(u.var.lb === Union{} && u.var.ub === Any)
                 return false
             end
-        elseif isa(ai, Const) && isa(ai.val, Type)
-            ai = ai.val
+        elseif (isa(ai, Const) && isa(ai.val, Type))
```
Comment (martinholters) [inline base/compiler/tfuncs.jl:1105]:
```suggestion
        elseif (isa(ai, Const) && isa(ai.val, Type)) || isconstType(ai)
```
or the change below is pointless.

**Review by martinholters — CHANGES_REQUESTED** [review]
Could use more tests of all the different code paths / conditions.

File: base/compiler/tfuncs.jl [hunk 2]
```diff
@@ -1067,6 +1069,20 @@ function _fieldtype_tfunc(@nospecialize(s), exact::Bool, @nospecialize(name))
 end
 add_tfunc(fieldtype, 2, 3, fieldtype_tfunc, 0)
 
+function valid_type_param(T)
```
Comment (JeffBezanson) [inline base/compiler/tfuncs.jl:1072]: Needs a more specific name to clarify the difference with `valid_tparam`. Maybe `all_instances_valid_tparams`?

File: base/compiler/tfuncs.jl [hunk 3]
```diff
@@ -1067,6 +1069,20 @@ function _fieldtype_tfunc(@nospecialize(s), exact::Bool, @nospecialize(name))
...
+    isvarargtype(T) && return false
```
Comment (JeffBezanson) [inline base/compiler/tfuncs.jl:1073]: I don't think this case is possible here.

Comment (JeffBezanson) [reply base/compiler/tfuncs.jl:1073]: This case is still wrong (it also was before, so it's not necessarily on this PR to fix it):
```
julia> Core.Compiler.apply_type_nothrow(Any[Core.Const(Ref),Core.Const(Vararg)],Type{<:Ref})
true
```

File: base/compiler/tfuncs.jl [hunk 4]
```diff
@@ -1086,22 +1102,38 @@ function apply_type_nothrow(argtypes::Array{Any, 1}, @nospecialize(rt))
     for i = 2:length(argtypes)
         isa(u, UnionAll) || return false
         ai = widenconditional(argtypes[i])
-        if ai ⊑ TypeVar
+        if ai ⊑ TypeVar || ai === DataType
```
Comment (JeffBezanson) [inline base/compiler/tfuncs.jl:1105]: Probably not true since `Vararg` isa DataType.

Comment (Keno) [reply base/compiler/tfuncs.jl:1105]: :(

File: base/compiler/tfuncs.jl (hunk 3 above)
Comment (Keno) [reply base/compiler/tfuncs.jl:1073]: > I don't think this case is possible here.

Maybe, but we may want to use this function from other places, so I thought it better to cover all the cases.

File: base/compiler/tfuncs.jl (hunk 2 above)
Comment (martinholters) [reply base/compiler/tfuncs.jl:1072]: `valid_tparam_type`?

File: base/compiler/tfuncs.jl (hunk 4 above)
Comment (Keno) [reply base/compiler/tfuncs.jl:1105]: @JeffBezanson and I talked about fixing this by making `Vararg` not a DataType, so we'll wait with this PR until @JeffBezanson gets to that change.

**Pull request merged**